### ISSUE TITLE
LGB model giving identical predictions

### ISSUE THREAD

**Issue by AdityaAshutosh** [microsoft/LightGBM#6879]
## Description
I am using LightGBM with regression objective to forecast product demands and was able to train a model using temporal features, lag features and rolling features. The dataset is fairly large and diverse in terms of pattern capturing capability. However, while performing predictions, the model is providing almost identical prediction for the target variable, despite of the input features fed for prediction being decently varied.

## Reproducible example

These are the parameters I am using. I have used lag features and rolling mean and rolling std features for a window of 7 days. 

```
lgb_params = {
                    'boosting_type': 'gbdt',
                    'objective': 'regression',
                    'metric': 'rmse',
                    'subsample': 0.8,
                    # 'subsample_freq': 1,
                    'learning_rate': 0.05,
                    'num_leaves': 63,
                    'min_data_in_leaf': 100,
                    'feature_fraction': 0.8,
                    # 'max_bin': 100,
                    'n_estimators': 2000,
                    'boost_from_average': True,
                    # 'verbose': -1,
                } 
```

This is the feature importance info:
name	imp
ItemNumber	17097
rolling_mean_1	16153
day_counter	10328
rolling_mean_2	9476
lag_1	8920
rolling_mean_3	8132
rolling_std_2	7532
Day	5673
lag_2	4897
day_of_week	3851
rolling_std_3	3833
lag_7	3009
rolling_mean_7	2827
rolling_mean_4	2782
lag_3	2724
lag_6	2677
Month	2331
rolling_mean_5	1907
rolling_std_4	1867
lag_5	1758
rolling_mean_6	1606
lag_4	1605
rolling_std_5	1172
rolling_std_7	827
rolling_std_6	611

How do I resolve the issue with these identical set of predictions?

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

That feature importance you've reported out rules out the most common source of this (the model having 0 informative splits).

Based just on the information you've provided, I suspect I understand what's happening.

I am *guessing* based only on their names that these features are to blame:

* `Day`: day of the month (integer from 1-31)
* `Month`: month of the year (integer from 1-12)
* `day_counter`: number of days since the earliest date in the dataset (integer starting at 1)

If I'm right about those descriptions, then I suspect that if you look in only the data you are calling `predict()` on, you'll find that for those specific features, the values are not actually "decently varied".

For example, if you provide 7 consecutive days' worth of data to `predict()`, that'll have at most 2 distinct values of `Month`, 7 distinct values of `Day`, and 7 distinct values of `day_counter`... that probably already excludes LOTs of paths through the trees in the model.

It also matters whether you've asked LightGBM to treat those features as categorical or continuous (you haven't told us).

The feature `day_counter` is especially problematic for a tree-based model (if I'm right about what its values and interpretation are). Recall that tree-based models like LightGBM, XGBoost, and CatBoost will find rules like "if `{feature} > {threshold}`, predict `p0`, else predict `p1`". If all of the values of `day_counter` in the data being passed to `predict()` are greater than `max(day_counter)` from the train data, then that could potentially eliminate many paths through the trees (because any `day_country < threshold` paths will never be followed).

LightGBM is often used with time-series data and for time-series forecasting, but it doesn't itself have any features specific to time-series data. If I'm right about those features, I recommend:

* drop feature `day_counter` completely
* treat `Day` and `Month` as categorical variables

If you're feeling unsure about how tree-based models change the requirements for working with timeseries data, you might also find it easier to use LightGBM via https://github.com/Nixtla/mlforecast.

**Comment by AdityaAshutosh** [comment]
Thankyou for such a quick turnaround. This seems to be very very specific and precise analysis and the assumptions you have made are correct.
Also, I have been treating the features as purely numerical values and not categorical, which might be a part of the problem. I will try out your recommendations and let you know if it works out

**Comment by jameslamb** [comment]
Alright, thanks. I'm going to add the `awaiting response` label, so this issue will automatically be closed in 30 days if we don't hear from you.

**Comment by AdityaAshutosh** [comment]
Hi @jameslamb . I dropped `day_counter` from the list of features. Also, I introduced `Day` , `Month` , `day_of_week` and `ItemNumber` as categorical variables. I also once introduced cyclic encoding features for `Day` and `Month`. However, there is no change in prediction behavior. 

Any other possible reason you suspect?

**Comment by jameslamb** [comment]
Some other possibilities:

* the data you're calling `predict()` on comes from a very different distribution than the training data
  - *check the minimum and maximum of every column, compare that to the training data... any that are very very different?*
  - *check the data types... maybe some features that were float64 in the training data are now float32 or int, and that loss of precision changes the path they take through the trees*
* the range of predicted values is very very small, and some numerical precision loss results in them all being the same
  - *check the range of the target variable from the training data and the range of predicted values*

Try calling `predict()` on the **training** data... do you see good performance metrics? And variability in the predictions roughly matching the variability in the target? If yes, we can be even more confident that the issue is that the data you're `predict()`-ing on comes from a different distribution than the training data, or at least a very small portion of the distribution from the training data.

Otherwise, it's very difficult to debug this given only the `params` dict and some feature importances. If you want more help, we'll need a reproducible example, which includes:

* how you're using LightGBM (Python, I'm assuming?)
* version of LightGBM
* how you installed it (exact commands)
* operating system

If possible, could you share a sample of the data you're predicting on, along with the model in text form (as produced by `Booster.save_model()`)?

**Comment by AdityaAshutosh** [comment]
Hi @jameslamb . I tried running predictions over training data and the performance was absolutely accurate. Also, there has been no issue like the loss of precision with the data types. I suspect the performance is hampered with the way I am handling feature engineering while building the prediction pipeline. The trained model is free of errors as per my assumption.

Just for a final thought, how does LightGBM deal with smaller variations in input features while making predictions. For example, for an input feature, if an increment of 1 unit is of fairly decisive deal, will LightGBM be able to recognize that variation in a better way as compared to other tree based models?

**Comment by jameslamb** [comment]
Great, glad you were able to find some changes in the feature engineering pipeline that helped to resolve the issue.

> *an increment of 1 unit*

Tree-based models like CatBoost, XGBoost, and LightGBM are just large collections of `if-else` statements, like `if feature_1 > 25.7, then predict 1.5 else predict 10.8`.

I think you'll find XGBoost's description of that helpful: https://xgboost.readthedocs.io/en/stable/tutorials/model.html

Given that, there isn't really a concept in these models of a "1 unit" change in an input feature having some exact impact on the predicted value, the way you might get with a linear model. For example, for continuous features the evaluation of those rules will be identical for any values larger than the maximum feature values observed during training.

If you want some behavior like that, you can try creating trees with linear models at the root, see https://lightgbm.readthedocs.io/en/latest/Parameters.html#linear_tree

Otherwise, you should think of it like this for tree-based ensembles.... if you want to make the model more sensitive to variation in the features, you should create a larger model. Where "larger" means some combination of "more trees" and "deeper trees".

In LightGBM, you could try that with parameter changes like this:

* increase `max_depth` and `num_leaves`
* decrease `min_data_in_bin`, `min_data_in_leaf`,  and/or `min_gain_to_split`

This is not an exhaustive list. LightGBM has many parameters to modify the learning process. See the documentation at https://lightgbm.readthedocs.io/en/latest/Parameters.html.
